src/runtime_src/tools/xclbinutil/SectionIPLayout.cxx

```diff
@@ -188,52 +236,89 @@ SectionIPLayout::marshalToJSON(char* _pDataSection,
+                   % index
+                   % getIPTypeStr((IP_TYPE)pHdr->m_ip_data[index].m_type)
+                   % getSubTypeStr((SUBTYPE)pHdr->m_ip_data[index].ps_kernel.m_subtype)
+                   % getFunctionalStr((FUNCTIONAL)pHdr->m_ip_data[index].ps_kernel.m_functional)
+                   % (unsigned int)pHdr->m_ip_data[index].ps_kernel.m_kernel_id
+                   % pHdr->m_ip_data[index].m_base_address
+                   % pHdr->m_ip_data[index].m_name);
+      } else {
       XUtil::TRACE(boost::format("[%d]: m_type: %s, properties: 0x%x, m_base_address: 0x%lx, m_name: '%s'")
                    % index
                    % getIPTypeStr((IP_TYPE)pHdr->m_ip_data[index].m_type)
                    % pHdr->m_ip_data[index].properties
                    % pHdr->m_ip_data[index].m_base_address
                    % pHdr->m_ip_data[index].m_name);
+      }
     }
 
     // Write out the entire structure
     XUtil::TRACE_BUF("ip_data", reinterpret_cast<const char*>(&(pHdr->m_ip_data[index])), sizeof(ip_data));
 
-    ip_data.put("m_type", getIPTypeStr((IP_TYPE)pHdr->m_ip_data[index].m_type).c_str());
+    ptIPEntry.put("m_type", getIPTypeStr((IP_TYPE)pHdr->m_ip_data[index].m_type).c_str());
+
+    switch ((IP_TYPE)pHdr->m_ip_data[index].m_type) {
+      case IP_MEM_DDR4:
+      case IP_MEM_HBM:
+      case IP_MEM_HBM_ECC:
+      {
+        ptIPEntry.put("m_index", (boost::format("%d") % (unsigned int)pHdr->m_ip_data[index].indices.m_index).str());
+        ptIPEntry.put("m_pc_index", (boost::format("%d") % (unsigned int)pHdr->m_ip_data[index].indices.m_pc_index).str());
+        break;
+      }
+      case IP_KERNEL:
+      {
+        ptIPEntry.put("m_int_enable", (boost::format("%d") % ((pHdr->m_ip_data[index].properties & ((uint32_t)IP_INT_ENABLE_MASK)))).str());
+        ptIPEntry.put("m_interrupt_id", (boost::format("%d") % (((pHdr->m_ip_data[index].properties & ((uint32_t)IP_INTERRUPT_ID_MASK)) >> IP_INTERRUPT_ID_SHIFT))).str());
+        std::string sIPControlType = getIPControlTypeStr((IP_CONTROL)((pHdr->m_ip_data[index].properties & ((uint32_t)IP_CONTROL_MASK)) >> IP_CONTROL_SHIFT));
+        ptIPEntry.put("m_ip_control", sIPControlType.c_str());
+        break;
+      }
+      case IP_PS_KERNEL:
+      {
+        // if m_subtype is ST_DPU (i.e. fixed ps kernel), display "m_subtype", "m_functional" and "m_kernel_id"
+        // else (non-fixed ps kernel), display "properties"
+        if ((SUBTYPE)pHdr->m_ip_data[index].ps_kernel.m_subtype == ST_DPU) { 
+          ptIPEntry.put("m_subtype", getSubTypeStr((SUBTYPE)pHdr->m_ip_data[index].ps_kernel.m_subtype));
+          ptIPEntry.put("m_functional", getFunctionalStr((FUNCTIONAL)pHdr->m_ip_data[index].ps_kernel.m_functional));
+          ptIPEntry.put("m_kernel_id", (boost::format("0x%x") % (unsigned int)pHdr->m_ip_data[index].ps_kernel.m_kernel_id).str());
+        } else {
```

Comment:
In this case, a break would be better.
```suggestion
            break;
        }
```

Reply:
IP_PS_KERNEL has two subtypes, the else block should handle the non-fixed ps kernel usecase, where we need to write out the "properties". If we put break here, the output ip layout won't have the "properties". Please correct me if I misunderstood.

Reply:
O.K.